packages/parser/src/ethereum/index.ts

```diff
@@ -0,0 +1,180 @@
+  private zrx: zrx.Parser
+
+  constructor(args: TransactionParserArgs) {
+    const provider = new ethers.providers.JsonRpcProvider(args.rpcUrl)
+
+    this.network = args.network ?? Network.Mainnet
+
+    this.thor = new thor.Parser({ network: this.network, midgardUrl: args.midgardUrl })
+    this.uniV2 = new uniV2.Parser({ network: this.network, provider })
+    this.zrx = new zrx.Parser()
+  }
+
+  // return any addresses that can be detected
+  getInternalAddress(inputData: string): string | undefined {
+    switch (getSigHash(inputData)) {
+      case this.thor.transferOutSigHash:
+        return this.thor.getInternalAddress(inputData)
+      case multiSig.SENDMULTISIG_SIG_HASH:
+        return multiSig.getInternalAddress(inputData)
+      default:
+        return
+    }
+  }
+
+  async parse(tx: Tx, address: string, internalTxs?: Array<InternalTx>): Promise<ParseTx> {
+    const sendAddress = tx.vin[0].addresses?.[0] ?? ''
+    const receiveAddress = tx.vout[0].addresses?.[0] ?? ''
+
+    let result: ParseTxSpecific | undefined
+    switch (receiveAddress) {
+      case zrx.PROXY_CONTRACT: {
+        result = this.zrx.parse(tx)
+        break
+      }
+      case this.thor.routerContract: {
+        result = await this.thor.parse(tx)
+        break
+      }
+      case uniV2.ROUTER_CONTRACT: {
+        result = await this.uniV2.parse(tx)
+        break
+      }
+    }
+
+    const pTx: ParseTx = {
+      address,
+      blockHash: tx.blockHash,
+      blockHeight: tx.blockHeight,
+      blockTime: tx.blockTime,
+      caip2: 'ETH',
+      trade: result?.trade,
+      transfers: result?.transfers ?? [],
+      txid: tx.txid,
+      value: tx.value,
+    }
+
+    if (address === sendAddress) {
+      // send amount
+      const sendValue = new BigNumber(tx.value)
+      if (!sendValue.isNaN() && sendValue.gt(0)) {
+        pTx.transfers = this.aggregateTransfer(
+          pTx.transfers,
+          TransferType.Send,
+          'ETH',
+          sendAddress,
+          receiveAddress,
+          sendValue.toString(10)
+        )
+      }
+
+      // network fee
+      const fees = new BigNumber(tx.fees ?? 0)
+      if (!fees.isNaN() && fees.gt(0)) {
+        pTx.fee = { caip19: 'ETH', value: fees.toString(10) }
+      }
+    }
+
+    if (address === receiveAddress) {
+      // receive amount
+      const receiveValue = new BigNumber(tx.value)
+      if (!receiveValue.isNaN() && receiveValue.gt(0)) {
+        pTx.transfers = this.aggregateTransfer(
+          pTx.transfers,
+          TransferType.Receive,
+          'ETH',
+          sendAddress,
+          receiveAddress,
+          receiveValue.toString(10)
+        )
+      }
+    }
+
+    tx.tokenTransfers?.forEach((transfer) => {
+      // FTX Token (FTT) name and symbol was set backwards on the ERC20 contract
+      if (transfer.token == '0x50D1c9771902476076eCFc8B2A83Ad6b9355a4c9') {
+        transfer.name = transfer.symbol
+        transfer.symbol = transfer.name
+      }
+
```

Comment:
This is mutating the `tx` object that's passed into the `parse` function. We should treat `tx` as immutable because the caller is probably not expecting their data to be mutated.

Reply:
I would agree typically, but this is an actual bug in the erc20 contract itself and we have had to add it to both wt and unchained. The intention is to swap the symbol and name values to be assigned correctly for this known contract only.